librubyfmt/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+#![deny(warnings, missing_copy_implementations)]
+use std::ffi::CString;
+#[macro_use]
+extern crate lazy_static;
+
+use std::io::{BufReader, Cursor, Write};
+use std::slice;
+
+#[global_allocator]
+static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
+
+pub type RawStatus = i64;
+
+mod breakable_entry;
+mod comment_block;
+mod de;
+mod delimiters;
+mod file_comments;
+mod format;
+mod intermediary;
+mod line_metadata;
+mod line_tokens;
+mod parser_state;
+mod render_queue_writer;
+mod ripper_tree_types;
+mod ruby;
+mod types;
+
+use file_comments::FileComments;
+use parser_state::ParserState;
+use ruby::VALUE;
+
+#[cfg(debug_assertions)]
+use log::debug;
+#[cfg(debug_assertions)]
+use simplelog::{Config, LevelFilter, TermLogger, TerminalMode};
+
+type RubyfmtResult<T = ()> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+extern "C" {
+    pub fn Init_ripper();
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct FormatBuffer {
+    pub bytes: *const libc::c_char,
+    pub count: i64,
+}
+
+impl FormatBuffer {
+    pub fn into_buf(self) -> &'static [u8] {
+        unsafe { slice::from_raw_parts(self.bytes as *const u8, self.count as usize) }
+    }
+
+    pub fn into_string(self) -> String {
+        unsafe {
+            let vec = Vec::from_raw_parts(
+                self.bytes as *mut u8,
+                self.count as usize,
+                self.count as usize,
+            );
+            String::from_utf8_unchecked(vec)
+        }
+    }
+}
```

Comment:
So *if* you were to try to keep this type, here's how I would write it to eliminate the unsoundness. I've noted the differences where they occur

```rust
#[repr(C)]
#[derive(Debug)] // This represents an owned, heap allocated pointer. To ensure it doesn't leak, it must have a `Drop` impl. It may also have methods that return similar owned types like `Vec`, `String`, or `Box`. It absolutely should not impl `Copy`, and a `Clone` impl would need to be manual and copy the memory being pointed to.

struct FormatBuffer {
    ptr: NonNull<u8>, // Encode as much as we can here. Using `u8`, C can cast the pointer all they want.
    len: usize, // Again, using the Rust type that's convenient. This is size_t or uintptr_t in C.
}

impl FormatBuffer {
    // Taking Into<Vec<u8>> since you're calling with String and Vec<u8>. `into_boxed_slice` copies, we can avoid
    // that if we keep the capacity around as well. Most importantly though, we provide a constructor which consumes
    // the input, leaving now room for one caller to accidentally use `.as_ptr()` instead of into_raw or similar. This 
    // struct should also live in its own file so its fields are private (C can of course see them, but we Rust callers
    // shouldn't).
    fn new<T: Into<Vec<u8>>>(bytes: T) -> Self {
        let bytes = bytes.into().into_boxed_slice();
        let len = bytes.len();
        Self {
            // safety: Box::into_raw never returns null. Use Box::into_raw_nonnull when stable
            ptr: unsafe { NonNull::new_unchecked(Box::into_raw(bytes) as _) },
            len,
        }
    }

    // Does not return 'static. Instead returns a proper lifetime.
    pub fn as_bytes(&self) -> &[u8] {
        // safety: We know we were constructed with a valid pointer
        unsafe { slice::from_raw_parts(self.bytes, self.len) }
    }

    // Returns `&str` instead of `String`. We own the pointer, we don't want to be returning
    // other types which own it.
    pub fn as_str(&self) -> &str {
        // Uses checked, we didn't validate UTF-8 on input. Even if we did, let's avoid unsafe
        str::from_utf8(self.as_bytes()).expect("Invalid UTF-8")
    }
}

impl Drop for FormatBuffer {
    fn drop(&mut self) {
        unsafe {
            let ptr = ptr::slice_from_raw_parts(self.ptr, self.len);
            drop(Box::from_raw(ptr));
        }
    }
}
```